### PR TITLE
Feature/parser

### PR DESCRIPTION
**Pull request by myrrlyn**
Clean up the parsers and start working on dealing with environment variables in text input

### CODE REVIEW

File: src/ast/mod.rs [hunk 1]
```diff
@@ -1,47 +1,124 @@
...
 pub use self::builtin::Builtin;
 pub use self::invoke::Invoke;
 
+#[derive(Debug, Clone)]
+pub struct ArgsIter<'a> {
+    text: &'a str
+}
+
+impl<'a> std::iter::Iterator for ArgsIter<'a> {
+    type Item = &'a str;
```
Comment (hawkw) [inline src/ast/mod.rs]: I think we're gonna want this to become an iterator of `Result<&'a str, SomeKindOfError>` eventually. Or, `next()` should halt the iteration and return none --- as it does currently --- but format the parse error first? Idk. But we need to stop swallowing errors soon.

File: src/ast/mod.rs [hunk 2]
```diff
@@ -1,47 +1,124 @@
...
     }
 }
 
-impl<'a> Parse<'a> for WithEnv<'a> {
-    type Error = String; // placeholder
-    fn parse_from(s: &'a str) -> Result<Self, ParseError<Self::Error>> {
-        // TODO(eliza): alex pls implement me
-        unimplemented!()
+#[cfg(test)]
+mod tests {
```
Comment (hawkw) [inline src/ast/mod.rs]: <3 <3 <3

File: src/env.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+/// The key=value pairs must occur at the front of the text. The keys must be
+/// bare words. They are broken by whitespace, and are not concerned with
+/// punctiation. It is valid, though foolish, to use a key `some"text`. There
+/// must be no whitespace between the key, the equals sign, and the beginning of
+/// the value. The value may be any text span as defined by the `parse::span`
+/// function: a single bare word, a single-quoted string, or a double-quoted
+/// string.
+#[derive(Clone, Debug)]
+crate struct EnvIter<'a> {
+    crate text: &'a str,
```
Comment (hawkw) [inline src/env.rs]: This doesn't appear to need to be `crate`?

File: src/ast/builtin.rs [hunk 4]
```diff
@@ -20,18 +20,12 @@ pub enum CdError {
...
-}
-
 // ===== impl Builtin =====
 
 impl<'a> Parse<'a> for Builtin<'a> {
     type Error = CdError;
-    fn parse_from(s: &'a str) -> Result<Self, ParseError<Self::Error>> {
-        let mut args = s.trim().split_whitespace();
+    fn parse_from(text: &'a str) -> Result<Self, ParseError<Self::Error>> {
+        let mut args = super::ArgsIter { text };
```
[the remark this answers is not in the document]
Comment (myrrlyn) [reply src/ast/builtin.rs]: it should I just didn't put one on it. I also want to split it into command and args so the sequence goes `env*, cmd, args*`

File: src/ast/mod.rs (hunk 1 above)
Comment (myrrlyn) [reply src/ast/mod.rs]: (extremely C voice) set a module-global error before returning None

Comment (hawkw) [reply src/ast/mod.rs]: no. fuck you.

File: src/ast/builtin.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (hawkw) [reply src/ast/builtin.rs]: ah, yeah, you might say something like `FOO="/home/eliza" cd $FOO` or something if you are an awful person, i guess

File: src/ast/mod.rs [hunk 5]
```diff
@@ -1,47 +1,124 @@
...
-    cmd: Cmd<'a>,
-    // TODO(eliza): what kind of fucked up iterator is the env going to be?
+    crate env: EnvIter<'a>,
+    crate cmd: Cmd<'a>,
 }
 
-impl<'a> Parse<'a> for Cmd<'a> {
+impl<'a> Parse<'a> for WithEnv<'a> {
     type Error = String; // placeholder
     fn parse_from(s: &'a str) -> Result<Self, ParseError<Self::Error>> {
```
Comment (hawkw) [inline src/ast/mod.rs]: oh, it occurs to me that POSIX sh also allows you to say `env FOO=bar BAZ=qworp whatever` and treats that as equivalent to `FOO=bar BAZ=qworp whatever` --- do we want to accept that form as well?

File: src/ast/builtin.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (myrrlyn) [reply src/ast/builtin.rs]: yeah, the actual evaluation of an input has to be able to refer to prior state, but more immediately the AST constructor can probably be chained child constructors

File: src/ast/mod.rs (hunk 5 above)
Comment (myrrlyn) [reply src/ast/mod.rs]: make env a builtin that either prints the environment (if no args) or skips and parses normally?

**Comment by myrrlyn** [comment]
yeah. I'm using `crate` visibility as a default so that I don't need to deal with visibility, and plan to constrict visibility as we get a more firm idea of the layout. I only want to make `pub` the items that are used in the mainloop, or that we think should actually be exported

**Pull request merged**